### PR TITLE
Jd/remove raw data

### LINKED ISSUE #359: The function `get_raw_data` is not type stable (see
These methods can be removed and change the requirement in PowerSimulations.jl

### PR DESCRIPTION
**Pull request by jd-lara**
This PR uses specific methods for each type. Closes #359

### CODE REVIEW

**Review by GabrielKS — APPROVED** [review]
Looks okay for now. At some point in the future I'd like to take another look at removing duplication here without introducing type instability, but that doesn't need to be a priority.

**Review by GabrielKS — CHANGES_REQUESTED** [review]
Actually, `get_raw_data_type` should probably be removed if you're removing `get_raw_data`

File: src/function_data.jl [hunk 1]
```diff
@@ -18,6 +18,29 @@ LinearFunctionData(proportional_term) = LinearFunctionData(proportional_term, 0.
 get_proportional_term(fd::LinearFunctionData) = fd.proportional_term
 get_constant_term(fd::LinearFunctionData) = fd.constant_term
 
+function _transform_linear_vector_for_hdf(data::Vector{LinearFunctionData})
+    transfd_data = Vector{NTuple{2, Float64}}(undef, length(data))
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/function_data.jl:22]: What's the problem with using Julia's syntax?

**Comment by jd-lara** [comment]
> Actually, `get_raw_data_type` should probably be removed if you're removing `get_raw_data`

Removing `get_raw_data_type` requires reimplementing the map as mentioned in #356 so we should defer that to another PR.

File: src/function_data.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (daniel-thom) [reply src/function_data.jl:22]: Isn't this the same thing but much simpler?
```
[(get_proportional_term(fd), get_constant_term(fd)) for (fd in data]
```
You could add the type if those aren't actually floats in `data`.

File: src/function_data.jl [hunk 2]
```diff
@@ -18,6 +18,29 @@ LinearFunctionData(proportional_term) = LinearFunctionData(proportional_term, 0.
...
+    return transfd_data
+end
+
+function transform_array_for_hdf(data::Vector{LinearFunctionData})
+    return transform_array_for_hdf(_transform_linear_vector_for_hdf(data))
+end
+
+function transform_array_for_hdf(
+    data::SortedDict{Dates.DateTime, Vector{LinearFunctionData}},
+)
```
Comment (daniel-thom) [inline src/function_data.jl:35]:
```
transform_array_for_hdf(SortedDict(k => _transform_linear_vector_for_hdf(v) for (k, v) in data])
```

Comment (jd-lara) [reply src/function_data.jl:35]: This won't allow for the use of `sizehint!`

File: src/function_data.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/function_data.jl:22]: Under which scenarios those won't be floats? I find the existing syntax much more explicit on the types

[the remark this answers is not in the document]
Comment (GabrielKS) [reply src/function_data.jl:22]: Readability-wise I much prefer the comprehension, I was assuming you were doing it this way for preallocation

File: src/function_data.jl (hunk 2 above)
Comment (daniel-thom) [reply src/function_data.jl:35]: sizehint shouldn't be necessary if you build like this.

File: src/function_data.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (daniel-thom) [reply src/function_data.jl:22]: My point is that this is heavy syntax to carry around when it isn't necessary. It adds maintenance for future developers. Or, if someone slightly changes one of the types, all of it has to be changed.

**Review by daniel-thom — APPROVED** [review]
My comments are only suggestions for simpler code.

**Pull request merged**